File: backend/app/schemas/material_flow.py
from __future__ import annotations
from datetime import datetime
from decimal import Decimal
from typing import List, Optional
from uuid import UUID
from pydantic import BaseModel, Field

from app.models.material_flow import (
    FlowType, FlowStatus, FlowMode, StageType, QualityStatus,
    MovementReason, ReservationStatus, TankStatus, ReconciliationStatus,
    MFAIAgentType, MFAIRecStatus,
)


# ── Stage ─────────────────────────────────────────────────────────────────────

class FlowStageCreate(BaseModel):
    stage_code: str
    stage_name: str
    stage_type: StageType
    plant_id: Optional[str] = None
    warehouse_id: Optional[UUID] = None
    location_id: Optional[UUID] = None
    quality_default_status: QualityStatus = QualityStatus.QUARANTINE
    sequence_no: int = 0
    active_flag: bool = True
    notes: Optional[str] = None


class FlowStageOut(FlowStageCreate):
    id: UUID
    warehouse_name: Optional[str] = None

    class Config:
        from_attributes = True


# ── Flow Transaction ──────────────────────────────────────────────────────────

class FlowLineCreate(BaseModel):
    line_no: int = 1
    item_type: str = "material"
    material_id: Optional[UUID] = None
    product_id: Optional[UUID] = None
    lot_id: Optional[UUID] = None
    batch_id: Optional[str] = None
    quantity: Decimal
    uom: str = "KG"
    source_warehouse_id: Optional[UUID] = None
    source_location_id: Optional[UUID] = None
    source_stage_id: Optional[UUID] = None
    destination_warehouse_id: Optional[UUID] = None
    destination_location_id: Optional[UUID] = None
    destination_stage_id: Optional[UUID] = None
    quality_status: QualityStatus = QualityStatus.RELEASED
    movement_reason: MovementReason = MovementReason.PRODUCTION_ISSUE
    reservation_flag: bool = False
    issue_flag: bool = False
    consumption_flag: bool = False
    return_flag: bool = False
    scrap_flag: bool = False
    rework_flag: bool = False
    reference_line_id: Optional[UUID] = None
    notes: Optional[str] = None


class FlowLineOut(FlowLineCreate):
    id: UUID
    transaction_id: UUID
    material_name: Optional[str] = None
    product_name: Optional[str] = None
    lot_number: Optional[str] = None
    source_stage_name: Optional[str] = None
    destination_stage_name: Optional[str] = None

    class Config:
        from_attributes = True


class FlowTransactionCreate(BaseModel):
    flow_type: FlowType
    flow_mode: FlowMode = FlowMode.ONE_STEP
    production_order_id: Optional[UUID] = None
    work_order_id: Optional[UUID] = None
    source_document_type: Optional[str] = None
    source_document_id: Optional[UUID] = None
    transaction_datetime: Optional[datetime] = None
    plant_id: Optional[str] = None
    notes: Optional[str] = None
    lines: List[FlowLineCreate] = Field(default_factory=list)


class FlowTransactionOut(BaseModel):
    id: UUID
    flow_no: str
    flow_type: FlowType
    flow_mode: FlowMode
    status: FlowStatus
    production_order_id: Optional[UUID] = None
    transaction_datetime: datetime
    plant_id: Optional[str] = None
    created_by: Optional[UUID] = None
    approved_by: Optional[UUID] = None
    notes: Optional[str] = None
    lines: List[FlowLineOut] = []

    class Config:
        from_attributes = True


class FlowTransactionApprove(BaseModel):
    notes: Optional[str] = None


# ── Reservation ───────────────────────────────────────────────────────────────

class ReservationLineCreate(BaseModel):
    line_no: int = 1
    item_type: str = "material"
    material_id: Optional[UUID] = None
    product_id: Optional[UUID] = None
    lot_id: Optional[UUID] = None
    warehouse_id: UUID
    quantity_required: Decimal
    uom: str = "KG"
    quality_status_required: QualityStatus = QualityStatus.RELEASED
    lot_fixed: bool = False
    notes: Optional[str] = None


class ReservationLineOut(ReservationLineCreate):
    id: UUID
    reservation_id: UUID
    quantity_reserved: Decimal = Decimal("0")
    quantity_issued: Decimal = Decimal("0")
    fefo_lot_candidate: Optional[str] = None
    material_name: Optional[str] = None
    lot_number: Optional[str] = None

    class Config:
        from_attributes = True


class ReservationCreate(BaseModel):
    production_order_id: UUID
    work_order_id: Optional[UUID] = None
    expiry_datetime: Optional[datetime] = None
    notes: Optional[str] = None
    lines: List[ReservationLineCreate] = Field(default_factory=list)


class ReservationOut(BaseModel):
    id: UUID
    reservation_no: str
    production_order_id: UUID
    status: ReservationStatus
    expiry_datetime: Optional[datetime] = None
    lines: List[ReservationLineOut] = []

    class Config:
        from_attributes = True


# ── Consumption ───────────────────────────────────────────────────────────────

class ConsumptionCreate(BaseModel):
    production_order_id: UUID
    work_order_id: Optional[UUID] = None
    item_type: str = "material"
    material_id: Optional[UUID] = None
    product_id: Optional[UUID] = None
    lot_id: Optional[UUID] = None
    quantity_consumed: Decimal
    quantity_returned: Decimal = Decimal("0")
    quantity_scrapped: Decimal = Decimal("0")
    uom: str = "KG"
    stage_id: Optional[UUID] = None
    notes: Optional[str] = None


class ConsumptionOut(ConsumptionCreate):
    id: UUID
    quantity_issued: Decimal
    quantity_variance: Decimal
    variance_pct: Optional[Decimal] = None
    bom_quantity: Optional[Decimal] = None
    consumed_at: Optional[datetime] = None
    material_name: Optional[str] = None

    class Config:
        from_attributes = True


# ── Prepared Lot ──────────────────────────────────────────────────────────────

class PreparedLotCreate(BaseModel):
    production_order_id: Optional[UUID] = None
    source_lot_id: Optional[UUID] = None
    material_id: UUID
    target_quantity: Decimal
    uom: str = "KG"
    tolerance_pct: Optional[Decimal] = None
    weighing_stage_id: Optional[UUID] = None
    notes: Optional[str] = None


class PreparedLotWeigh(BaseModel):
    actual_quantity: Decimal
    operator_id: Optional[UUID] = None
    notes: Optional[str] = None


class PreparedLotOut(PreparedLotCreate):
    id: UUID
    prepared_lot_no: str
    actual_quantity: Optional[Decimal] = None
    over_weigh: bool = False
    under_weigh: bool = False
    quality_status: QualityStatus = QualityStatus.RELEASED
    consumed: bool = False
    weighed_at: Optional[datetime] = None
    material_name: Optional[str] = None

    class Config:
        from_attributes = True


# ── Tank Occupancy ────────────────────────────────────────────────────────────

class TankCreate(BaseModel):
    tank_code: str
    tank_name: str
    stage_id: Optional[UUID] = None
    warehouse_id: Optional[UUID] = None
    capacity_max: Optional[Decimal] = None
    uom: str = "KG"
    notes: Optional[str] = None


class TankUpdate(BaseModel):
    product_id: Optional[UUID] = None
    material_id: Optional[UUID] = None
    lot_id: Optional[UUID] = None
    batch_id: Optional[str] = None
    production_order_id: Optional[UUID] = None
    quantity_current: Optional[Decimal] = None
    quantity_reserved: Optional[Decimal] = None
    quality_status: Optional[QualityStatus] = None
    tank_status: Optional[TankStatus] = None
    notes: Optional[str] = None


class TankOut(BaseModel):
    id: UUID
    tank_code: str
    tank_name: str
    stage_id: Optional[UUID] = None
    warehouse_id: Optional[UUID] = None
    capacity_max: Optional[Decimal] = None
    uom: str
    product_id: Optional[UUID] = None
    material_id: Optional[UUID] = None
    lot_id: Optional[UUID] = None
    batch_id: Optional[str] = None
    production_order_id: Optional[UUID] = None
    quantity_current: Decimal
    quantity_reserved: Decimal
    quality_status: QualityStatus
    tank_status: TankStatus
    fill_datetime: Optional[datetime] = None
    release_datetime: Optional[datetime] = None
    notes: Optional[str] = None
    product_name: Optional[str] = None
    material_name: Optional[str] = None
    fill_pct: Optional[Decimal] = None

    class Config:
        from_attributes = True


# ── Reconciliation ────────────────────────────────────────────────────────────

class ReconciliationLineOut(BaseModel):
    id: UUID
    line_no: int
    item_type: str
    material_id: Optional[UUID] = None
    product_id: Optional[UUID] = None
    bom_quantity: Optional[Decimal] = None
    issued_quantity: Decimal
    consumed_quantity: Decimal
    returned_quantity: Decimal
    scrapped_quantity: Decimal
    variance_quantity: Decimal
    variance_pct: Optional[Decimal] = None
    uom: str
    variance_reason: Optional[str] = None
    cost_impact: Optional[Decimal] = None
    material_name: Optional[str] = None

    class Config:
        from_attributes = True


class ReconciliationOut(BaseModel):
    id: UUID
    reconciliation_no: str
    production_order_id: UUID
    status: ReconciliationStatus
    total_issued: Decimal
    total_consumed: Decimal
    total_returned: Decimal
    total_scrapped: Decimal
    total_variance: Decimal
    variance_pct: Optional[Decimal] = None
    output_planned: Optional[Decimal] = None
    output_actual: Optional[Decimal] = None
    output_variance: Optional[Decimal] = None
    variance_reason: Optional[str] = None
    approved_by: Optional[UUID] = None
    approved_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None
    lines: List[ReconciliationLineOut] = []

    class Config:
        from_attributes = True


class ReconciliationApprove(BaseModel):
    variance_reason: Optional[str] = None
    notes: Optional[str] = None


# ── AI Recommendations ────────────────────────────────────────────────────────

class MFAIRecOut(BaseModel):
    id: UUID
    agent_type: MFAIAgentType
    status: MFAIRecStatus
    production_order_id: Optional[UUID] = None
    title: str
    description: str
    severity: str
    confidence_score: Optional[Decimal] = None
    estimated_impact: Optional[str] = None
    actioned_by: Optional[UUID] = None
    actioned_at: Optional[datetime] = None
    action_notes: Optional[str] = None

    class Config:
        from_attributes = True


class MFAIRecAction(BaseModel):
    status: MFAIRecStatus
    notes: Optional[str] = None


# ── Dashboard ─────────────────────────────────────────────────────────────────

class MFDashboard(BaseModel):
    total_flows_today: int = 0
    total_issued_today: Decimal = Decimal("0")
    active_reservations: int = 0
    open_reconciliations: int = 0
    tanks_in_use: int = 0
    tanks_qc_hold: int = 0
    pending_ai_recs: int = 0
    total_wip_quantity: Decimal = Decimal("0")
    recent_flows: List[FlowTransactionOut] = []
    ai_recommendations: List[MFAIRecOut] = []


# ── Reports ───────────────────────────────────────────────────────────────────

class FlowHistoryFilter(BaseModel):
    production_order_id: Optional[UUID] = None
    material_id: Optional[UUID] = None
    lot_id: Optional[UUID] = None
    warehouse_id: Optional[UUID] = None
    stage_id: Optional[UUID] = None
    flow_type: Optional[FlowType] = None
    date_from: Optional[datetime] = None
    date_to: Optional[datetime] = None


class VarianceReportRow(BaseModel):
    production_order_id: UUID
    order_no: str
    material_id: Optional[UUID] = None
    material_name: Optional[str] = None
    bom_quantity: Optional[Decimal] = None
    issued_quantity: Decimal
    consumed_quantity: Decimal
    returned_quantity: Decimal
    variance_quantity: Decimal
    variance_pct: Optional[Decimal] = None
    uom: str

File: frontend/src/lib/materialFlow.ts
import apiClient from "@/lib/api";

// ── Enums ────────────────────────────────────────────────────────────────────

export type FlowType =
  | "RAW_ISSUE" | "PKG_ISSUE" | "CONSUMABLE_ISSUE"
  | "WEIGHING_TRANSFER" | "STAGE_TRANSFER"
  | "INTERMEDIATE_CREATE" | "BULK_TRANSFER"
  | "TANK_FILL" | "TANK_TRANSFER" | "LINE_TRANSFER"
  | "FG_RECEIPT" | "RETURN" | "REVERSAL"
  | "SCRAP" | "REWORK" | "RECONCILIATION";

export type FlowStatus = "DRAFT" | "PENDING" | "CONFIRMED" | "POSTED" | "REVERSED" | "CANCELLED";

export type FlowMode = "ONE_STEP" | "TWO_STEP" | "THREE_STEP" | "MULTI_STAGE";

export type StageType =
  | "RAW_WAREHOUSE" | "QUARANTINE_RAW" | "WEIGHING_ROOM" | "STAGED_LINE"
  | "MIXING" | "PREMIX" | "INTERMEDIATE_HOLD" | "BULK_TANK" | "QC_HOLD_TANK"
  | "RELEASED_BULK" | "FILLING_STAGING" | "PACKAGING_STAGING"
  | "SECONDARY_PACKING" | "PALLETIZATION" | "FG_QUARANTINE" | "FG_RELEASED"
  | "REWORK_HOLDING" | "SCRAP_HOLDING" | "TRANSIT";

export type QualityStatus =
  | "QUARANTINE" | "PENDING_INSPECTION" | "RELEASED" | "QC_HOLD"
  | "BLOCKED" | "REJECTED" | "REWORK_PENDING";

export type ReservationStatus = "ACTIVE" | "PARTIALLY_ISSUED" | "FULLY_ISSUED" | "EXPIRED" | "CANCELLED";

export type TankStatus = "EMPTY" | "FILLING" | "FULL" | "IN_USE" | "QC_HOLD" | "CLEANING" | "OUT_OF_SERVICE";

export type ReconciliationStatus = "OPEN" | "IN_PROGRESS" | "PENDING_APPROVAL" | "APPROVED" | "CLOSED";

export type MFAIAgentType = "VARIANCE_ANALYZER" | "FLOW_OPTIMIZER" | "RISK_MONITOR";

export type MFAIRecStatus = "PENDING" | "ACCEPTED" | "REJECTED" | "APPLIED";

// ── Types ────────────────────────────────────────────────────────────────────

export interface FlowStage {
  id: string;
  stage_code: string;
  stage_name: string;
  stage_type: StageType;
  plant_id?: string;
  warehouse_id?: string;
  warehouse_name?: string;
  quality_default_status: QualityStatus;
  sequence_no: number;
  active_flag: boolean;
  notes?: string;
}

export interface FlowLine {
  id: string;
  transaction_id: string;
  line_no: number;
  item_type: string;
  material_id?: string;
  product_id?: string;
  lot_id?: string;
  batch_id?: string;
  quantity: number;
  uom: string;
  source_warehouse_id?: string;
  destination_warehouse_id?: string;
  source_stage_id?: string;
  destination_stage_id?: string;
  source_stage_name?: string;
  destination_stage_name?: string;
  quality_status: QualityStatus;
  movement_reason: string;
  issue_flag: boolean;
  return_flag: boolean;
  scrap_flag: boolean;
  material_name?: string;
  product_name?: string;
  lot_number?: string;
}

export interface FlowTransaction {
  id: string;
  flow_no: string;
  flow_type: FlowType;
  flow_mode: FlowMode;
  status: FlowStatus;
  production_order_id?: string;
  transaction_datetime: string;
  plant_id?: string;
  notes?: string;
  lines: FlowLine[];
}

export interface FlowTransactionCreate {
  flow_type: FlowType;
  flow_mode?: FlowMode;
  production_order_id?: string;
  notes?: string;
  lines: Partial<FlowLine>[];
}

export interface Reservation {
  id: string;
  reservation_no: string;
  production_order_id: string;
  status: ReservationStatus;
  expiry_datetime?: string;
  lines: ReservationLine[];
}

export interface ReservationLine {
  id: string;
  reservation_id: string;
  line_no: number;
  item_type: string;
  material_id?: string;
  warehouse_id: string;
  quantity_required: number;
  quantity_reserved: number;
  quantity_issued: number;
  uom: string;
  material_name?: string;
  lot_number?: string;
}

export interface Consumption {
  id: string;
  production_order_id: string;
  material_id?: string;
  quantity_issued: number;
  quantity_consumed: number;
  quantity_returned: number;
  quantity_scrapped: number;
  quantity_variance: number;
  variance_pct?: number;
  uom: string;
  consumed_at?: string;
  material_name?: string;
}

export interface PreparedLot {
  id: string;
  prepared_lot_no: string;
  production_order_id?: string;
  material_id: string;
  target_quantity: number;
  actual_quantity?: number;
  uom: string;
  over_weigh: boolean;
  under_weigh: boolean;
  quality_status: QualityStatus;
  consumed: boolean;
  weighed_at?: string;
  material_name?: string;
}

export interface TankOccupancy {
  id: string;
  tank_code: string;
  tank_name: string;
  stage_id?: string;
  warehouse_id?: string;
  capacity_max?: number;
  uom: string;
  product_id?: string;
  material_id?: string;
  lot_id?: string;
  production_order_id?: string;
  quantity_current: number;
  quantity_reserved: number;
  quality_status: QualityStatus;
  tank_status: TankStatus;
  fill_datetime?: string;
  release_datetime?: string;
  product_name?: string;
  material_name?: string;
  fill_pct?: number;
}

export interface Reconciliation {
  id: string;
  reconciliation_no: string;
  production_order_id: string;
  status: ReconciliationStatus;
  total_issued: number;
  total_consumed: number;
  total_returned: number;
  total_scrapped: number;
  total_variance: number;
  variance_pct?: number;
  output_planned?: number;
  output_actual?: number;
  output_variance?: number;
  variance_reason?: string;
  approved_at?: string;
  closed_at?: string;
  lines: ReconciliationLine[];
}

export interface ReconciliationLine {
  id: string;
  line_no: number;
  material_id?: string;
  bom_quantity?: number;
  issued_quantity: number;
  consumed_quantity: number;
  returned_quantity: number;
  scrapped_quantity: number;
  variance_quantity: number;
  variance_pct?: number;
  uom: string;
  cost_impact?: number;
  material_name?: string;
}

export interface MFAIRec {
  id: string;
  agent_type: MFAIAgentType;
  status: MFAIRecStatus;
  production_order_id?: string;
  title: string;
  description: string;
  severity: string;
  confidence_score?: number;
  estimated_impact?: string;
  actioned_at?: string;
}

export interface MFDashboard {
  total_flows_today: number;
  total_issued_today: number;
  active_reservations: number;
  open_reconciliations: number;
  tanks_in_use: number;
  tanks_qc_hold: number;
  pending_ai_recs: number;
  total_wip_quantity: number;
  recent_flows: FlowTransaction[];
  ai_recommendations: MFAIRec[];
}

// ── Label / Color Maps ────────────────────────────────────────────────────────

export const FLOW_TYPE_LABELS: Record<FlowType, string> = {
  RAW_ISSUE: "Raw Material Issue",
  PKG_ISSUE: "Packaging Issue",
  CONSUMABLE_ISSUE: "Consumable Issue",
  WEIGHING_TRANSFER: "Weighing Transfer",
  STAGE_TRANSFER: "Stage Transfer",
  INTERMEDIATE_CREATE: "Intermediate Create",
  BULK_TRANSFER: "Bulk Transfer",
  TANK_FILL: "Tank Fill",
  TANK_TRANSFER: "Tank Transfer",
  LINE_TRANSFER: "Line Transfer",
  FG_RECEIPT: "FG Receipt",
  RETURN: "Return",
  REVERSAL: "Reversal",
  SCRAP: "Scrap",
  REWORK: "Rework",
  RECONCILIATION: "Reconciliation",
};

export const FLOW_STATUS_COLORS: Record<FlowStatus, string> = {
  DRAFT: "bg-gray-100 text-gray-700",
  PENDING: "bg-yellow-100 text-yellow-700",
  CONFIRMED: "bg-blue-100 text-blue-700",
  POSTED: "bg-green-100 text-green-700",
  REVERSED: "bg-orange-100 text-orange-700",
  CANCELLED: "bg-red-100 text-red-700",
};

export const QUALITY_STATUS_COLORS: Record<QualityStatus, string> = {
  QUARANTINE: "bg-yellow-100 text-yellow-700",
  PENDING_INSPECTION: "bg-blue-100 text-blue-700",
  RELEASED: "bg-green-100 text-green-700",
  QC_HOLD: "bg-orange-100 text-orange-700",
  BLOCKED: "bg-red-100 text-red-800",
  REJECTED: "bg-red-200 text-red-900",
  REWORK_PENDING: "bg-purple-100 text-purple-700",
};

export const TANK_STATUS_COLORS: Record<TankStatus, string> = {
  EMPTY: "bg-gray-100 text-gray-600",
  FILLING: "bg-blue-100 text-blue-700",
  FULL: "bg-green-100 text-green-700",
  IN_USE: "bg-emerald-100 text-emerald-700",
  QC_HOLD: "bg-orange-100 text-orange-700",
  CLEANING: "bg-cyan-100 text-cyan-700",
  OUT_OF_SERVICE: "bg-red-100 text-red-700",
};

export const AI_AGENT_LABELS: Record<MFAIAgentType, string> = {
  VARIANCE_ANALYZER: "Variance Analyzer",
  FLOW_OPTIMIZER: "Flow Optimizer",
  RISK_MONITOR: "Risk Monitor",
};

export const SEVERITY_COLORS: Record<string, string> = {
  critical: "bg-red-100 text-red-800",
  high: "bg-orange-100 text-orange-800",
  medium: "bg-yellow-100 text-yellow-800",
  low: "bg-green-100 text-green-800",
};

// ── API Client ────────────────────────────────────────────────────────────────

const BASE = "/api/v1/material-flow";

export const mfApi = {
  getDashboard: () => apiClient.get<MFDashboard>(`${BASE}/dashboard`),

  // Stages
  listStages: () => apiClient.get<FlowStage[]>(`${BASE}/stages`),
  createStage: (d: Partial<FlowStage>) => apiClient.post<FlowStage>(`${BASE}/stages`, d),

  // Transactions
  listTransactions: (params?: Record<string, string>) =>
    apiClient.get<FlowTransaction[]>(`${BASE}/transactions`, { params }),
  getTransaction: (id: string) => apiClient.get<FlowTransaction>(`${BASE}/transactions/${id}`),
  createTransaction: (d: FlowTransactionCreate) => apiClient.post<FlowTransaction>(`${BASE}/transactions`, d),
  confirmTransaction: (id: string) => apiClient.post<FlowTransaction>(`${BASE}/transactions/${id}/confirm`, {}),
  reverseTransaction: (id: string) => apiClient.post<FlowTransaction>(`${BASE}/transactions/${id}/reverse`, {}),

  // Issues / Returns / Transfers / FG
  issueMaterials: (d: FlowTransactionCreate) => apiClient.post<FlowTransaction>(`${BASE}/issues`, d),
  returnMaterials: (d: FlowTransactionCreate) => apiClient.post<FlowTransaction>(`${BASE}/returns`, d),
  transferMaterials: (d: FlowTransactionCreate) => apiClient.post<FlowTransaction>(`${BASE}/transfers`, d),
  fgReceipt: (d: FlowTransactionCreate) => apiClient.post<FlowTransaction>(`${BASE}/fg-receipts`, d),

  // Reservations
  listReservations: (params?: Record<string, string>) =>
    apiClient.get<Reservation[]>(`${BASE}/reservations`, { params }),
  getReservation: (id: string) => apiClient.get<Reservation>(`${BASE}/reservations/${id}`),
  createReservation: (d: object) => apiClient.post<Reservation>(`${BASE}/reservations`, d),
  cancelReservation: (id: string) => apiClient.post<Reservation>(`${BASE}/reservations/${id}/cancel`, {}),

  // Consumptions
  recordConsumption: (d: object) => apiClient.post<Consumption>(`${BASE}/consumptions`, d),
  listConsumptions: (orderId: string) => apiClient.get<Consumption[]>(`${BASE}/consumptions/${orderId}`),

  // Prepared lots
  listPreparedLots: (params?: Record<string, string>) =>
    apiClient.get<PreparedLot[]>(`${BASE}/prepared-lots`, { params }),
  createPreparedLot: (d: object) => apiClient.post<PreparedLot>(`${BASE}/prepared-lots`, d),
  recordWeigh: (id: string, d: object) => apiClient.post<PreparedLot>(`${BASE}/prepared-lots/${id}/weigh`, d),

  // Tanks
  listTanks: (params?: Record<string, string>) => apiClient.get<TankOccupancy[]>(`${BASE}/tanks`, { params }),
  getTank: (id: string) => apiClient.get<TankOccupancy>(`${BASE}/tanks/${id}`),
  createTank: (d: object) => apiClient.post<TankOccupancy>(`${BASE}/tanks`, d),
  updateTank: (id: string, d: object) => apiClient.patch<TankOccupancy>(`${BASE}/tanks/${id}`, d),

  // Reconciliation
  getReconciliation: (orderId: string) => apiClient.get<Reconciliation>(`${BASE}/reconciliation/${orderId}`),
  closeReconciliation: (orderId: string, d?: object) =>
    apiClient.post<Reconciliation>(`${BASE}/reconciliation/${orderId}/close`, d || {}),

  // History
  getHistory: (params?: Record<string, string>) =>
    apiClient.get<FlowTransaction[]>(`${BASE}/history`, { params }),

  // AI
  runAI: (orderId?: string) =>
    apiClient.post<MFAIRec[]>(`${BASE}/ai/run`, {}, { params: orderId ? { production_order_id: orderId } : {} }),
  listAIRecs: (params?: Record<string, string>) =>
    apiClient.get<MFAIRec[]>(`${BASE}/ai/recommendations`, { params }),
  actionAIRec: (id: string, status: MFAIRecStatus, notes?: string) =>
    apiClient.post<MFAIRec>(`${BASE}/ai/recommendations/${id}/action`, { status, notes }),
};
